Create rule S6379[terraform]: Enabling Azure resource-specific admin accounts is security-sensitive (#573)
SonarSource/rspec: rules/S6379/terraform/rule.adoc

----

== Compliant Solution

For https://docs.microsoft.com/en-us/azure/batch/nodes-and-pools#pools[Azure Batch Pools]:

----
resource "azurerm_batch_pool" "example" {
  name = "example"

  start_task {
    command_line = "echo 'Hello World'"
    max_task_retry_count = 1
    wait_for_success = true

    user_identity {
      auto_user {
        elevation_level = "NonAdmin"
        scope = "Task"
      }
    }
  }
}
----

For https://azure.microsoft.com/en-us/services/container-registry/[Azure Container Registries]:

----
resource "azurerm_container_registry" "exemple" {
  name = "example"
  admin_enabled = false
}
----

== See

* https://owasp.org/Top10/A01_2021-Broken_Access_Control/[OWASP Top 10 2021 Category A1] - Broken Access Control
* https://owasp.org/www-project-top-ten/2017/A5_2017-Broken_Access_Control[OWASP Top 10 2017 Category A5] - Broken Access Control
* https://cwe.mitre.org/data/definitions/284.html[MITRE, CWE-284] - Improper Access Control

ifdef::env-github,rspecator-view[]

'''
== Implementation Specification
(visible only on this page)

=== Message

Make sure that enabling an administrative account or administrative permissions is safe here.

=== Highlighting

> * For Azure Batch Pools, Highlight `start_task`.

Reviewer: ```suggestion
* For Azure Batch Pools, Highlight `elevation_level = "NonAdmin"` or `start_task`.
```

> * For Azure Batch Pools, Highlight ``elevation_level = "Admin"`` or ``start_task``.

Reviewer: ```suggestion
* For Azure Batch Pools, Highlight ``elevation_level = "Admin"``.
```

> * For Azure Batch Pools, Highlight `elevation_level = "Admin"` or `start_task`.
> * For Azure container Registries, highlight `admin_enabled = true` or the resource.
> 

Reviewer: ```suggestion
* For Azure Batch Pools, Highlight ``elevation_level = "Admin"`` or ``start_task``.
* For Azure container Registries, highlight ``admin_enabled = true`` or the resource.

```

> * For Azure container Registries, highlight `admin_enabled = false` or the resource.

Reviewer 2: I guess the highlighting should be `elevation_level = "Admin"` and `admin_enabled = true`.
For Batch Pools,  in which case should `start_task` be highlighted? It's not clear to me as the ticket says that `elevation_level`defaults to `NonAdmin`.

Reviewer: Yes, absolutely. This is a mistake

Reviewer: ```suggestion
* For Azure Batch Pools, Highlight `elevation_level = "Admin"` or `start_task`.
* For Azure container Registries, highlight `admin_enabled = true` or the resource.
```

> Enabling administrative permissions can reduce an organization's ability to protect itself against account or service account thefts.

Reviewer 2: ```suggestion
Enabling Azure resource-specific admin accounts can reduce an organization's ability to protect itself against account or service account thefts.
```

Full Administrator permissions fail to correctly separate duties and create potentially critical attack vectors on the impacted resources.

In case of abuse of elevated permissions, both the data on which impacted resources operate and their access traceability are at risk.

== Ask Yourself Whether

* This Azure resource is essential for the information system infrastructure.
* This Azure resource is essential for mission-critical functions.
* Compliance policies require this resource to disable its administrative accounts or permissions.

There is a risk if you answered yes to any of those questions.

== Recommended Secure Coding Practices

Disable the administrative accounts or permissions in this Azure resource.

== Sensitive Code Example

For https://docs.microsoft.com/en-us/azure/batch/nodes-and-pools#pools[Azure Batch Pools]:

----

> resource "azurerm_batch_pool" "sensitive_pool" {

Reviewer 2: ```suggestion
resource "azurerm_batch_pool" "example" {
```

  name = "sensitive"

  start_task {
    command_line = "echo 'Hello World'"
    max_task_retry_count = 1
    wait_for_success = true

  user_identity {
    auto_user {
      elevation_level = "Admin" # Sensitive
      scope = "Task"
      }
    }
  }
}
----

For https://azure.microsoft.com/en-us/services/container-registry/[Azure Container Registries]:

----

> resource "azurerm_container_registry" "sensitive_registry" {

Reviewer 2: ```suggestion
resource "azurerm_container_registry" "example" {
```

  name = "sensitive"
  admin_enabled = true # Sensitive
}
----

== Compliant Solution

For https://docs.microsoft.com/en-us/azure/batch/nodes-and-pools#pools[Azure Batch Pools]:

----

> resource "azurerm_batch_pool" "sensitive_pool" {

Reviewer 2: ```suggestion
resource "azurerm_batch_pool" "example" {
```

>   name = "sensitive"

Reviewer 2: ```suggestion
  name = "example"
```

  start_task {
    command_line = "echo 'Hello World'"
    max_task_retry_count = 1
    wait_for_success = true

  user_identity {
    auto_user {
      elevation_level = "NonAdmin"
      scope = "Task"
      }
    }
  }
}
----

For https://azure.microsoft.com/en-us/services/container-registry/[Azure Container Registries]:

----

> resource "azurerm_container_registry" "sensitive_registry" {
>   name = "sensitive"
>   admin_enabled = false
> }

Reviewer 2: ```suggestion
resource "azurerm_container_registry" "exemple" {
  name = "example"
  admin_enabled = false
}
```

----

== See

> * https://owasp.org/Top10/A01_2021-Broken_Access_Control/[OWASP Top 10 2021 Category A1] - Broken Access Control

Reviewer: ```suggestion
* https://owasp.org/Top10/A01_2021-Broken_Access_Control/[OWASP Top 10 2021 Category A1] - Broken Access Control
* https://owasp.org/www-project-top-ten/2017/A5_2017-Broken_Access_Control[OWASP Top 10 2017 Category A5 - Broken Access Control
```

> * https://owasp.org/Top10/A01_2021-Broken_Access_Control/[OWASP Top 10 2021 Category A1] - Broken Access Control

Reviewer 2: Add OWASP Top 10 2017 category too
